rest/bytes_read_public_api_test.go

```diff
@@ -0,0 +1,475 @@
+
+	_, ok = base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(newStat))
+	require.True(t, ok)
+}
+
+func TestBytesReadPOSTSession(t *testing.T) {
+	rt := NewRestTester(t, nil)
+	defer rt.Close()
+
+	rt.CreateUser("alice", []string{"ABC"})
+
+	// create session
+	input := `{"name":"alice","password":"letmein"}`
+	inputBytes := []byte(input)
+	resp := rt.SendUserRequest(http.MethodPost, "/{{.db}}/_session", input, "alice")
+	RequireStatus(t, resp, http.StatusOK)
+
+	// assert the stat is increased by the correct amount
+	_, ok := base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
+	require.True(t, ok)
+
+	newStat := len(inputBytes) * 2
+	// now try failed auth
+	resp = rt.SendUserRequest(http.MethodPost, "/{{.db}}/_session", input, "bob")
+	RequireStatus(t, resp, http.StatusUnauthorized)
+
+	_, ok = base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(newStat))
+	require.True(t, ok)
+}
+
+func TestBytesReadAuthFailed(t *testing.T) {
+	rt := NewRestTester(t, nil)
+	defer rt.Close()
+
+	// create a user with different password to the default one
+	resp := rt.SendAdminRequest(http.MethodPut, "/{{.db}}/_user/alice", GetUserPayload(t, "alice", "pass", "", rt.GetSingleTestDatabaseCollection(), []string{"ABC"}, nil))
+	RequireStatus(t, resp, http.StatusCreated)
+
+	// make a request that will fail on auth
+	input := `{"docs": [{"_id": "bulk1", "n": 1}, {"_id": "bulk2", "n": 2}, {"_id": "_local/bulk3", "n": 3}]}`
+	inputBytes := []byte(input)
+	resp = rt.SendUserRequest("POST", "/{{.keyspace}}/_bulk_docs", input, "alice")
+	RequireStatus(t, resp, http.StatusUnauthorized)
+
+	// assert the stat has still increased by the bytes of the body passed into request
+	_, ok := base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
+	require.True(t, ok)
+
+}
+
+func TestBytesReadGzipRequest(t *testing.T) {
+	// Need default collection as request below doesn't work with {{.keyspace}}
+	rt := NewRestTesterDefaultCollection(t, &RestTesterConfig{
+		GuestEnabled: true,
+	})
+	defer rt.Close()
+
+	rt.CreateUser("alice", []string{"ABC"})
+	input := `{"channel":["ABC"]}`
+
+	var buf bytes.Buffer
+	gz := gzip.NewWriter(&buf)
+	_, err := gz.Write([]byte(input))
+	require.NoError(t, err)
+	err = gz.Close()
+	require.NoError(t, err)
+
+	inputBytes := []byte(input)
+
+	// {{.keyspace}} isn't supported so use default collection
+	rq, err := http.NewRequest("PUT", "/db/doc", &buf)
+	require.NoError(t, err)
+	rq.Header.Set("Content-Encoding", "gzip")
+	resp := rt.Send(rq)
+	RequireStatus(t, resp, http.StatusCreated)
+
+	_, ok := base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
```

Comment:
We're counting the decompressed request body here - which seems wrong in context of calculating actual bandwidth used.

Might (hopefully) just be a case of re-ordering the io.Reader wrappers:
E.g. `counter(gzip(req.Body)) = counts uncompressed` vs. `gzip(counter(req.Body)) = counts compressed`

Reply:
Going with approximation due to the underlying encoded byte array being private and the gzip utility decompressing before reading.